# Covid-19 Data Analysis

# Introduction

### - Coronaviruses are a large family of viruses which may cause illness in animals or humans. In humans, several coronaviruses are known to cause respiratory infections ranging from the common cold to more severe diseases such as Middle East Respiratory Syndrome (MERS) and Severe Acute Respiratory Syndrome (SARS). The most recently discovered coronavirus causes coronavirus disease COVID-19 - World Health Organization

### - The number of cases and deaths realted to COVID-19 are increasing daily. We need to know where the diseases is headed in terms of numbers in order to control the pandemic from destroying more lives. 

## 01 - Import Libraries

## 02 - Import Data

## 03 - Details about the columns

## 04 - Wrangling data
  <ul>
<li>4.1 Renaming values</li>
<li>4.2 Dropping unnecessary Columns</li>
  </ul> 

## 05 - Consistensy checks
  <ul>
<li>5.1 Checking for mixed data types</li>
 <li>5.2 Checking for missing values</li>
 <li>5.3 Checking for duplicates</li>
  </ul>

## 06 - Descriptive Analysis

## 07 - Export data

# 01 - Import libraries

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02 - Import data

In [3]:
# Create a Master path

path = r'/Users/jagpreetbatth/Desktop/Career Foundry/Python/COVID-19 Analysis'

In [4]:
# Check Path Creation

path

'/Users/jagpreetbatth/Desktop/Career Foundry/Python/COVID-19 Analysis'

In [6]:
# Import Covid-19 Dataset

df= pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'covid_19_clean_complete.csv'))

In [8]:
# Print head

df.head()

Province/State Country/Region       Lat       Long        Date  Confirmed  \
0            NaN    Afghanistan  33.93911  67.709953  2020-01-22          0   
1            NaN        Albania  41.15330  20.168300  2020-01-22          0   
2            NaN        Algeria  28.03390   1.659600  2020-01-22          0   
3            NaN        Andorra  42.50630   1.521800  2020-01-22          0   
4            NaN         Angola -11.20270  17.873900  2020-01-22          0   

   Deaths  Recovered  Active             WHO Region  
0       0          0       0  Eastern Mediterranean  
1       0          0       0                 Europe  
2       0          0       0                 Africa  
3       0          0       0                 Europe  
4       0          0       0                 Africa

In [12]:
# Print tail

df.tail()

Province/State         Country/Region        Lat       Long        Date  \
49063            NaN  Sao Tome and Principe   0.186400   6.613100  2020-07-27   
49064            NaN                  Yemen  15.552727  48.516388  2020-07-27   
49065            NaN                Comoros -11.645500  43.333300  2020-07-27   
49066            NaN             Tajikistan  38.861000  71.276100  2020-07-27   
49067            NaN                Lesotho -29.610000  28.233600  2020-07-27   

       Confirmed  Deaths  Recovered  Active             WHO Region  
49063        865      14        734     117                 Africa  
49064       1691     483        833     375  Eastern Mediterranean  
49065        354       7        328      19                 Africa  
49066       7235      60       6028    1147                 Europe  
49067        505      12        128     365                 Africa

In [13]:
# Print Shape

df.shape

(49068, 10)

# 03 - Details about Columns

In [9]:
df.columns

Index(['Province/State', 'Country/Region', 'Lat', 'Long', 'Date', 'Confirmed',
       'Deaths', 'Recovered', 'Active', 'WHO Region'],
      dtype='object')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49068 entries, 0 to 49067
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Province/State  14664 non-null  object 
 1   Country/Region  49068 non-null  object 
 2   Lat             49068 non-null  float64
 3   Long            49068 non-null  float64
 4   Date            49068 non-null  object 
 5   Confirmed       49068 non-null  int64  
 6   Deaths          49068 non-null  int64  
 7   Recovered       49068 non-null  int64  
 8   Active          49068 non-null  int64  
 9   WHO Region      49068 non-null  object 
dtypes: float64(2), int64(4), object(4)
memory usage: 3.7+ MB


### There are 5 numerical columns ('Lat','Long','Confirmed','Deaths','Recovered') , 2 catagorical columns ('Province/State' & 'Country/Region' ) and 1 datetime column ('Date')

In [14]:
# Study Unique Values

df.describe(include = 'object')

Province/State Country/Region        Date WHO Region
count                          14664          49068       49068      49068
unique                            78            187         188          6
top     Australian Capital Territory          China  2020-01-22     Europe
freq                             188           6204         261      15040

### Going in detail, there are 187 unique countries and 78 unique province/states enlisted in the dataset

In [15]:
# Time Period

a = df.Date.value_counts().sort_index()
print('The first date is:',a.index[0])
print('The last date is:',a.index[-1])

The first date is: 2020-01-22
The last date is: 2020-07-27


### The dataset starts from 22nd January to 27th July 2020

# 04 - Wrangling Data 

## 4.1 - Rename Values

In [20]:
# Renaming the coulmns for easy usage

df.rename(columns={'Province/State':'State','Country/Region':'Country'},inplace=True)

In [21]:
# Confirm re-name changes

df.columns

Index(['State', 'Country', 'Lat', 'Long', 'Date', 'Confirmed', 'Deaths',
       'Recovered', 'Active', 'WHO Region'],
      dtype='object')

## 4.2 - Dropping unnecessary Values

### Since the column "State" includes a lot of NaN, and it's needless for this for this analysis, column dropped.

In [22]:
# Dropping "State" column from df

df = df.drop(columns = ['State'])

In [23]:
# Confirm Drop of "State"

df.head()

Country       Lat       Long        Date  Confirmed  Deaths  Recovered  \
0  Afghanistan  33.93911  67.709953  2020-01-22          0       0          0   
1      Albania  41.15330  20.168300  2020-01-22          0       0          0   
2      Algeria  28.03390   1.659600  2020-01-22          0       0          0   
3      Andorra  42.50630   1.521800  2020-01-22          0       0          0   
4       Angola -11.20270  17.873900  2020-01-22          0       0          0   

   Active             WHO Region  
0       0  Eastern Mediterranean  
1       0                 Europe  
2       0                 Africa  
3       0                 Europe  
4       0                 Africa

In [24]:
# Print shape after dropping column

df.shape

(49068, 9)

# 05 - Data Consistency Checks

### 5.1 Checking for mixed data types

In [25]:
#Checking for columns with mixed data types

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

#### No mixed Data types resulted

### 5.2 Checking for missing values

In [26]:
# Checking for missing values in df.

df.isnull().sum()

Country       0
Lat           0
Long          0
Date          0
Confirmed     0
Deaths        0
Recovered     0
Active        0
WHO Region    0
dtype: int64

#### No missing values resulted

### 5.3 Checking for duplicates

In [27]:
#Looking for duplicates in df (creating a new subset containing only duplicates)

df_dups = df[df.duplicated()]

In [28]:
#Print df_dups

df_dups

Empty DataFrame
Columns: [Country, Lat, Long, Date, Confirmed, Deaths, Recovered, Active, WHO Region]
Index: []

#### No Duplicates noted

# 06 - Descriptive Analysis

In [29]:
# Print basic statistics post consistency check.

df.describe()

Lat          Long     Confirmed         Deaths     Recovered  \
count  49068.000000  49068.000000  4.906800e+04   49068.000000  4.906800e+04   
mean      21.433730     23.528236  1.688490e+04     884.179160  7.915713e+03   
std       24.950320     70.442740  1.273002e+05    6313.584411  5.480092e+04   
min      -51.796300   -135.000000  0.000000e+00       0.000000  0.000000e+00   
25%        7.873054    -15.310100  4.000000e+00       0.000000  0.000000e+00   
50%       23.634500     21.745300  1.680000e+02       2.000000  2.900000e+01   
75%       41.204380     80.771797  1.518250e+03      30.000000  6.660000e+02   
max       71.706900    178.065000  4.290259e+06  148011.000000  1.846641e+06   

             Active  
count  4.906800e+04  
mean   8.085012e+03  
std    7.625890e+04  
min   -1.400000e+01  
25%    0.000000e+00  
50%    2.600000e+01  
75%    6.060000e+02  
max    2.816444e+06

#### Note the minimum of active cases is -14, we cannot have a negative number of cases that are active. This minimum negative value and others like it come from the equation: [Active Case = confirmed - deaths - recovered]. For example, if the confirmed cases have a value lower than the amount recovered, you will get a negative value. In simpler terms, if a person has recovered, tested negative for the COVID-19 virus, they are essentially no longer active, decreasing the likelihood of transmissibility and transmission of the virus to others, making them a part of the recovered pool.

#### Solution: Since the negative values only make up 18 of the 49,068 rows, a very small portion of our data, and for the purposes of our final analysis focusing on confrimed cases, deaths and recovered only, we will replace the negative values with NaN.

In [38]:
df['Active'].replace(-14,np.nan, inplace=True)
df['Active'].replace(-13,np.nan, inplace=True)
df['Active'].replace(-10,np.nan, inplace=True)
df['Active'].replace(-6,np.nan, inplace=True)
df['Active'].replace(-4,np.nan, inplace=True)
df['Active'].replace(-3,np.nan, inplace=True)
df['Active'].replace(-2,np.nan, inplace=True)
df['Active'].replace(-1,np.nan, inplace=True)

In [39]:
# Check the minimum value after replacing Negative numbers

df.min()

Country       Afghanistan
Lat              -51.7963
Long               -135.0
Date           2020-01-22
Confirmed               0
Deaths                  0
Recovered               0
Active                0.0
WHO Region         Africa
dtype: object

#### Active cases minimum is correctly at 0

In [40]:
# Re-Check basic statistics after replacing Negative numbers

df.describe()

Lat          Long     Confirmed         Deaths     Recovered  \
count  49068.000000  49068.000000  4.906800e+04   49068.000000  4.906800e+04   
mean      21.433730     23.528236  1.688490e+04     884.179160  7.915713e+03   
std       24.950320     70.442740  1.273002e+05    6313.584411  5.480092e+04   
min      -51.796300   -135.000000  0.000000e+00       0.000000  0.000000e+00   
25%        7.873054    -15.310100  4.000000e+00       0.000000  0.000000e+00   
50%       23.634500     21.745300  1.680000e+02       2.000000  2.900000e+01   
75%       41.204380     80.771797  1.518250e+03      30.000000  6.660000e+02   
max       71.706900    178.065000  4.290259e+06  148011.000000  1.846641e+06   

             Active  
count  4.905000e+04  
mean   8.087981e+03  
std    7.627274e+04  
min    0.000000e+00  
25%    0.000000e+00  
50%    2.600000e+01  
75%    6.060000e+02  
max    2.816444e+06

#### Minumum values active cases are at 0, meaning there are no "negative" Active cases values, which is not possible. 

#  07 - Export Data

In [41]:
#Export df to Prepared Data

df.to_csv(os.path.join(path, 'Data','Prepared Data', 'COVID-19_analysis_clean.csv'))